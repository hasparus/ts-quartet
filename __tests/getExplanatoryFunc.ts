export const getExplanatoryFunc = (
  validValue: string,
  explanation: string
) => () => ({
  check: (v: any) => `${v} === ${JSON.stringify(validValue)}`,
  not: (v: any) => `${v} !== ${JSON.stringify(validValue)}`,
  handleError: (v: any, ctx: any) =>
    `${ctx}.explanations.push(${JSON.stringify(explanation)})`
});
