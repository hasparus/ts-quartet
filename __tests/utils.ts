import { getDescription } from "./getDescription";

export function snapshot(validator: any) {
  const description = getDescription(validator);

  expect(description).toMatchSnapshot();
}
export function getExplanation(
  validator: { (arg0: any): void; explanations: any },
  value: any
) {
  validator(value);
  return validator.explanations;
}
export function isValid(validator: (arg0: any) => any, value: any) {
  const valid = validator(value);
  if (!valid) {
    console.log(JSON.stringify(value, null, 2));
  }
  expect(valid).toBe(true);
}
export function isNotValid(validator: (arg0: any) => any, value: any) {
  const valid = validator(value);
  if (valid) {
    console.log(JSON.stringify(value, null, 2));
  }
  expect(valid).toBe(false);
}

export function tables(validator: any, valids: any[], invalids: any[]) {
  for (const valid of valids) {
    isValid(validator, valid);
  }
  for (const invalid of invalids) {
    isNotValid(validator, invalid);
  }
}
