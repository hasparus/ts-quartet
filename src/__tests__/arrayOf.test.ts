import { v } from "../index";
import { snapshot } from "./utils";
import { getExplanatoryFunc } from "./getExplanatoryFunc";

describe("arrayOf", () => {
  test("constant schema", () => {
    snapshot(v.arrayOf(1));
    snapshot(v.arrayOf("1"));
    snapshot(v.arrayOf("true"));
    snapshot(v.arrayOf(Symbol.for("name")));
  });
  test("function", () => {
    snapshot(v.arrayOf(getExplanatoryFunc("A", "Is not A")));
  });
});
