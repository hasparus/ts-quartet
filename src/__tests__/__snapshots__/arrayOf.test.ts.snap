// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`arrayOf constant schema 1`] = `
Object {
  "_": "function validator(value) {
    if (!value || !Array.isArray(value)) return false
    for (let i = 0; i < value.length; i++) {
      const elem = value[i]
      if (elem !== 1) return false
    }
    return true
  }",
  "explanations": Array [],
}
`;

exports[`arrayOf constant schema 2`] = `
Object {
  "_": "function validator(value) {
    if (!value || !Array.isArray(value)) return false
    for (let i = 0; i < value.length; i++) {
      const elem = value[i]
      if (elem !== \\"1\\") return false
    }
    return true
  }",
  "explanations": Array [],
}
`;

exports[`arrayOf constant schema 3`] = `
Object {
  "_": "function validator(value) {
    if (!value || !Array.isArray(value)) return false
    for (let i = 0; i < value.length; i++) {
      const elem = value[i]
      if (elem !== 'true') return false
    }
    return true
  }",
  "explanations": Array [],
}
`;

exports[`arrayOf constant schema 4`] = `
Object {
  "_": "function validator(value) {
    if (!value || !Array.isArray(value)) return false
    for (let i = 0; i < value.length; i++) {
      const elem = value[i]
      if (elem !== validator['symbol-element-0']) return false
    }
    return true
  }",
  "explanations": Array [],
  "symbol-element-0": Symbol(name),
}
`;

exports[`arrayOf function 1`] = `
Object {
  "_": "function validator(value) {
    if (!value || !Array.isArray(value)) return false
    for (let i = 0; i < value.length; i++) {
      const elem = value[i]
      if (elem !== \\"A\\") {
        validator.explanations.push(\\"Is not A\\")
        return false
      }
    }
    return true
  }",
  "explanations": Array [],
}
`;
